File: src/main/java/com/example/skyr/note/sync/NotesSyncDao.java
package com.example.skyr.note.sync;

import com.example.skyr.Dao;
import lombok.RequiredArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public final class NotesSyncDao extends Dao {
    private final JdbcTemplate jdbcTemplate;

    public NotesSync readWithLock() {
        final var result = jdbcTemplate.query(
                "SELECT sync_id, sync_time FROM notes_sync FOR UPDATE",
                (resultSet, rowNum) -> NotesSyncFactory.create(resultSet)
        );
        return singleResult(result);
    }

    public void update(final NotesSync notesSync) {
        jdbcTemplate.update(
                "UPDATE notes_sync SET sync_id = ?, sync_time = ?",
                notesSync.syncId().toString(),
                timestamp(notesSync.syncTime()).toString()
        );
    }
}

File: src/main/kotlin/com/example/skyr/note/sync/NotesSyncDao.kt
package com.example.skyr.note.sync

import com.example.skyr.Dao
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Component

@Component
class NotesSyncDao(private val jdbcTemplate: JdbcTemplate) : Dao() {

    companion object {
        const val NOTES_SYNC = "notes_sync"
    }

    fun read(writeLock: Boolean = false): NotesSync? {
        val result = jdbcTemplate.query(
            "SELECT ${NotesSync.SYNC_ID}, ${NotesSync.SYNC_TIME} FROM $NOTES_SYNC" + forUpdate(writeLock)
        ) { resultSet, _ -> notesSync(resultSet) }
        return resolveSingleResult(result)
    }

    fun update(notesSync: NotesSync) {
        jdbcTemplate.update(
            "UPDATE $NOTES_SYNC SET ${NotesSync.SYNC_ID} = ?, ${NotesSync.SYNC_TIME} = ?",
            notesSync.syncId.toString(),
            timestamp(notesSync.syncTime).toString()
        )
    }
}
